tests/test_product.py

```diff
@@ -0,0 +1,51 @@
+import pytest
+from buy_together_app import models
+
+QR = "Q5o76MbdiNbXprNEnHfpcGWFp1CMF8XY"
+NAME = "Banana"
+DESCRIPTION = "Good!"
+
+
+@pytest.fixture
+def product0():
+    return models.Product(qr_code=QR, product_name=NAME, description=DESCRIPTION)
+
+
+@pytest.fixture
+def saved_product0(product0):
+    product0.save()
+    return product0
+
+
+class TestProductModel:
+    def test_new_product(self, product0):
+        assert product0.qr_code == QR
+        assert product0.product_name == NAME
+        assert product0.description == DESCRIPTION
+
+    @pytest.mark.django_db()
+    def test_save_product(self, saved_product0):
+        assert saved_product0 in models.Product.objects.all()
+
+    @pytest.mark.django_db()
+    def test_delete_product(self, saved_product0):
+        assert saved_product0 in models.Product.objects.all()
+        saved_product0.delete()
+        assert saved_product0 not in models.Product.objects.all()
+
+    @pytest.mark.django_db()
+    def test_filter_by_qr(self, saved_product0):
+        assert [saved_product0] == list(models.Product.filter_qr(saved_product0.qr_code))
+
+    @pytest.mark.django_db()
+    def test_filter_by_product_name(self, saved_product0):
+        assert saved_product0 in list(models.Product.filter_name(saved_product0.product_name))
```

Comment:
this will not fail even if the function returns all products for some reason. I think the approach in `test_filter_by_qr` is more accurate.

Reply:
We can have multiple product names that is why,qr is unique

Reply:
How do you know whether you got back all products or just the ones you expected?

Reply:
because the fixture product name is not empty,if you are searching for a "" product name then you will know that you will get all the products but if you have given a name then you know that you will ge the result

Reply:
saved_product0  is a product with qr,name,desc if we check saved_product in the list it means we need a one product in the list that has the same qr and name and desc and since qr is unique we have it

Reply:
> because the fixture product name is not empty,if you are searching for a "" product name then you will know that you will get all the products but if you have given a name then you know that you will ge the result

You're assuming here the functionality which you need to test, which is empty string should return one thing and a non-empty string should return another.

Why not using `==`?

Reply:
Sorry. I didn't get what you're saying.

Reply:
I'm not saying that `product_name` might be an empty string. I'm saying that the filter functionality for treating empty string in one way and non-empty string in another way is what you're trying to test here.

This is your code now:

```python
    @staticmethod
    def filter_name(nm):
        if nm:
            return Product.objects.filter(product_name=nm)
        return Product.objects.all()
```

Suppose that tomorrow someone introduces a new feature and changes it

```python
    @staticmethod
    def filter_name(nm):
        if nm and something_that_returns_false_due_to_a_bug(nm):
            return Product.objects.filter(product_name=nm)
        return Product.objects.all()
```

Your test will still pass, as the product will be on the list.

Your test assumes something about the code it validates which it doesn't have to assume, and it can potentially make it miss regressions in your code in the future.

Reply:
Updated